Team forming feature v0.1 release

wickathou opened the pull request:
# V0.1

## Features:
- [x] Participants can browse teams and apply to roles listed by the team
- [x] Participants with a team can see candidate applications to a team role and accept or decline their application
- [x] Team members and owners can view and edit their team's profile

### Future updates:
- Allow participants to edit or withdraw their application to a team
- Clean up unused variables, packages and styling from updated and new files created
- Abstract elements rendered on pages into independent components
- Merge @mrprotocoll socialLinks component

wickathou committed:
Removed console logs from team controller
Added team roles creation and deletion logic
Added prop isTeamRoles to Select component to handled teamRole creation
Updated ParticipantPreview to use the property profilePicture instead of avatar
Started to implement Redux and backend for team formation logic on the Apply and candidates page
Created TeamCreateEditForm component to abstract logic from the team creation page
Added conditional verification to teamRoles properties to avoid typeErrors
Modified TeamRole schema to include candidates that apply to a role

wickathou committed:
Updated TeamRole schema to only store information about the roles and not about the candidates
Added candidateApplyToTeam method to add users to the candidate pool of a team
Added candidates field to the team model to track applications
Added getTeamById to enable team querying by id and candidateApplyToTeam handlers for candidate applications
Updated Apply component to send data to the team model when applicant submits application
Updated candidate component to render candidates only if they exist for a team
Updated teams page with dashboard actions for updating a seleted team and passing team role data as a prop to Apply component
Added new candidateApplyToTeam action to the dashboard actions
Added new candidateApplyToTeam axios service to the TeamsServices
Created new Candidate schema

wickathou committed:
Updated team controller to createNewTeam without members property
Added attachUserApplicant method to determine if a user has applied to a team
Updated team model to require code to be unique and generate shortids whenever a new team is created
Updated team route for getTeamsForEvent to received the user sub id as param
Updated candidate card to render candidate data when team candidates exist, along with the roles they applied for
Updated teamCard to visually indicate if a user has applied to a team already
Updated Participant preview to use avatar instead of profilePicture
Updated component Apply handleApply to pass more information about the user sending an application to a team
Updated candidates page to render candidates only if they exist
Modified candidate schema to include firstName, lastName and avatar
Modified TeamRole schema to allow open application

wickathou committed:
Added headline property to the candidate schema

wickathou committed:
Added avatar property to the recruitment controller createRecruitmentProfile
> Created function to generate unique codes for teams
> Created getUserProfileById method for user-profile controller to fetch userProfiles using their userId
> Added new getUserProfileById route
> Modified CandidateCard to receive and render dynamic data passed from props
> Created Participant Profile component
> Created ProfileInfo component to be used inside the Profile component
Created ProfileSide component to render inside the Profile component
Modified candidate page to fetch user profiles and their applications, plus rendering data conditionally based on the page state
Refactored variable slug inside of the team profile page
Updated recruitment participant profile view to use the new Participant Profile component to render user data
Added getCandidateProfileById to fetch user profiles using their Id
Created new dashboard actionType GET_CANDIDATE_PROFILE
Added getUserProfileById method to the userProfileService

wickathou committed:
Created userProfileUtils to format userProfiles returned by the backend
Added acceptCandidateToTeam method to the team controller to add a candidate as a team member
Fixed issue on the team model where the team code would be regenerated when updates are saved in a team
Added new route to team routes to handle accepting a candidate to a team
Modified CandidateCard to handle when team members accept a candidate to join their team
Modified Participant ProfileInfo to avoid errors when the user data passed is null or undefined
Modified Participant ProfileSide, ProfileTop and Profile to handled cases when prop data is passed as null or undefined
Created RecruitementFavorites component to separate favorites from the Profile component
Modified objToArr inside dataModifiers utility to handle cases when the object passed is not an object type
Modified candidates page to render data on the same page conditionally based on the state
Modified profile team page to hide team profile when the user is editing or creating a team
Added acceptCandidateToTeam method to the dashboard actions
Added acceptCandidateToTeam method to the teams services

wickathou committed:
Added new method declineCandidateToTeam to team controller, routes, dashboard actions and team service to remove applicant from a team candidate pool

wickathou committed:
Update team acceptCandidateToTeam controller method to remove users from other teams when the candidate is accepted in a team
Updated team getTeamsForEvent controller method to return status of users as applicant conditionally
Updated multiple methods and functions using the same name getUserProfileById to getUserPublicProfileById to make clear that the method returns enriched data about the user for profile rendering
Created a component filter to add filtering functionality to team related pages
Updated the team creation/edit form component style

wickathou committed:
Removed commented code from Filter component

wickathou committed:
Improved team controller candidateApplyToTeam method to prevent that users apply if they are part of the team they are trying to apply to or if they are part of a team
Updated CandidateCard handleApply to handleAccept
Updated min height of the teamCard
Added multiple TODO indicators for future debugging and refactoring
Updated filter component to be usable on any page
Updated team profile with prop enableActions to make the TeamProfile to be rendered on more contexts than the team profile page
Modified TeamRoles to receive onRoleClick prop to pass updates to the state
Added snackbar and filters to the candidates page
Added filter to the teams page
Added error indicator to the dashboard actions candidateApplyToTeam method
Added conditional verification to the finalistVoting property inside the showPages dashboard selector

wickathou committed:
Added a new todo for the teams page

wickathou committed:
Uncommented code that allows participants to be added to favorites on recruitment profile view page

wickathou committed:
Added new schema to validate team inputs during creation and editing
Removed team-bk backup for team creation
Fixed issue on team candidates page that would fail to load if a candidate does not have a team

wickathou committed:
Fixed issue on the team controller createNewTeam method that prevented the creation of a team if the teamRoles is empty
Added new props to teamCard component and teams page to prevent users from applying to a role if they already have a team
Fixed team apply component not showing errors on validation
Added frontend validation to the TeamCreateEditForm component
Added social rendering for team profiles
Created regexUtils to keep all regex validation utilities in one file

samurotko commented:
**Review fail:**

- [ ] fix apply team
- [ ] fix discord and telegram links

Full review here: https://www.notion.so/junctionhq/Team-forming-review-d8d401393fd4497b8fc8132c1e56a492?pvs=4

Otherwise good for now @wickathou.

wickathou committed:
Fixed bugs on team forming feature teams page and profile page

wickathou committed:
Removed unused code from team controller candidateApplyToTeam
Tried additional ways to pass team candidate validation which continues failing

wickathou committed:
Fixed issue that prevented users from applying with an open application when applying to a team

samurotko commented:
Review ok! ✅ 

Things to fix later:
https://www.notion.so/junctionhq/Team-forming-review-d8d401393fd4497b8fc8132c1e56a492?pvs=4

Merging to junction platform 2.0 branch and publishing to TEST environment
